doc/source/data/doc_code/key_concepts.py

```diff
@@ -1,32 +1,57 @@
 # flake8: noqa
 
 # fmt: off
-# __resource_allocation_begin__
+# __resource_allocation_1_begin__
 import ray
+from ray import tune
+
+# This Dataset workload will use spare cluster resources for execution.
+def objective(*args):
+    ray.data.range(10).show()
+
+# Create a cluster with 4 CPU slots available.
+ray.init(num_cpus=4)
+
+# This runs, since Tune schedules one trial on 1 CPU, leaving 3 spare CPUs in the
+# cluster for Dataset execution. However, deadlock can occur if you set num_samples=4,
```

Comment:
Can the same problem occur if Tune and Dataset compete for resources within the PG like they compete within cluster?

Reply:
Technically no, assuming your PG is properly configured. If you configure it wrong then yes :)